### PR TITLE
128-bit SIMD vector primitive type

### PR DESCRIPTION
**Pull request by TheNumbat**
Adds a new primitive type (`vec128`) representing a boxed 128-bit SIMD vector.
Currently, values of this type are opaque, so can only be passed among functions/stored in boxed data structures.
They may also be passed unboxed to external C calls that expect `__m128` parameters/returns.

Features to add in future PRs:
- primitive type per specific packed type + typed intrinsics for SSE operations
- otherlibs module providing the user facing API
- bigstring / bigarray / load from float array support
- vec128array of unboxed vec128s (for float32 / float64 only)
- vec128 layout + unboxed 128 bit types

Later, the same for 256-bit and 512-bit operations.

Current omissions:
- No arm64 support
- No upstream compiler backend support
- Does not assure vectors stored in memory are 16-byte aligned, except stack spills/stack arguments for extcalls

### CODE REVIEW

**Comment by mshinwell** [comment]
I'm starting to wonder if we should in fact not update `ocaml/` for Cmm language changes here.  Let's discuss.

**Review by goldfirere — APPROVED** [review]
Only files owned by me

**Review by gretay-js — CHANGES_REQUESTED** [review]
I reviewed the files I own. Only have small comments (note a copy-paste bug in cmm_helper).

I'm not confident about register allocator changes. They seem fine to me but I am not sure if there are any other places to update. @xclerc can you please review it.

When adding ADDPD it needs to be handled in selection and reload slightly differently from Iadd on float and int registers, because there is no two-address form, as far as I can see.

At some point we should also rememeber to add a line to backend/amd64/NOTES to document the instruction set level supported for SIMD.

File: backend/cmm.mli [hunk 1]
```diff
@@ -258,6 +265,7 @@ type expression =
     Cconst_int of int * Debuginfo.t
   | Cconst_natint of nativeint * Debuginfo.t
   | Cconst_float of float * Debuginfo.t
+  | Cconst_vec128 of int64 * int64 * Debuginfo.t
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/cmm.mli:268]: Changed to a record and added a comment

File: backend/cmm_helpers.ml [hunk 2]
```diff
@@ -2737,17 +2793,17 @@ let tuplify_function arity return =
 
 let max_arity_optimized = 15
 
+let ints_per_8b = if Arch.size_int = 4 then 2 else 1
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/cmm_helpers.ml:2796]: Changed

File: ocaml/asmcomp/cmmgen.ml [hunk 3]
```diff
@@ -1282,7 +1290,7 @@ and transl_let env str (layout : Lambda.layout) id exp transl_body =
          there may be constant closures inside that need lifting out. *)
       let _cbody : expression = transl_body env in
       cexp
-  | Punboxed_float | Punboxed_int _ -> begin
+  | Punboxed_float | Punboxed_int _ | Punboxed_vector _ -> begin
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply ocaml/asmcomp/cmmgen.ml:1293]: I think this case is correct for the unboxed vector case, but I guess it should be an error for consistency

File: backend/amd64/emit.mlp [hunk 4]
```diff
@@ -1112,9 +1161,9 @@ let emit_instr fallthrough i =
       I.movq a0 (res i 0);
       I.neg (res i 0)
   | Lop(Inegf) ->
-      I.xorpd (mem64_rip OWORD (emit_symbol "caml_negf_mask")) (res i 0)
+      I.xorpd (mem64_rip VEC128 (emit_symbol "caml_negf_mask")) (res i 0)
   | Lop(Iabsf) ->
-      I.andpd (mem64_rip OWORD (emit_symbol "caml_absf_mask")) (res i 0)
+      I.andpd (mem64_rip VEC128 (emit_symbol "caml_absf_mask")) (res i 0)
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/emit.mlp:1166]: OWORD was already only used for loading a 128-bit constant into an XMM register, so I thought it'd be more consistent to re-use it as VEC128. I could change it to XMMWORD if you think that'd be better.

File: backend/x86_masm.ml [hunk 5]
```diff
@@ -19,8 +19,8 @@ open X86_proc
 let bprintf = Printf.bprintf
 
 let string_of_datatype = function
+  | VEC128 -> "VEC128"
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/x86_masm.ml:22]: Yeah it should be XMMWORD

File: backend/cmm_helpers.ml [hunk 6]
```diff
@@ -81,6 +81,11 @@ let float_header = block_header Obj.double_tag (size_float / size_addr)
 let float_local_header =
   local_block_header Obj.double_tag (size_float / size_addr)
 
+let boxedvec128_header = block_header Obj.abstract_tag (16 / size_addr)
+
+let boxedvec128_local_header =
+  local_block_header Obj.abstract_tag (16 / size_addr)
+
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/cmm_helpers.ml:88]: Float needs a special tag for distinguishing flat float arrays - we're not planning on doing that for vectors, so they can all use abstract tag

File: backend/cmm_helpers.ml [hunk 7]
```diff
@@ -699,6 +713,40 @@ let rec unbox_float dbg =
...
+let box_vector dbg vi m c =
+  Cop (Calloc m, [alloc_boxedvector_header vi m dbg; c], dbg)
+
+let rec unbox_vector dbg vi =
+  let load = match vi with Primitive.Pvec128 -> Onetwentyeight in
+  map_tail ~kind:Any (function
+    | Cop (Calloc _, [Cconst_natint (hdr, _); c], _)
+      when Nativeint.equal hdr float_header
+           || Nativeint.equal hdr float_local_header ->
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/cmm_helpers.ml:726]: Oops, yes

File: backend/cmm_helpers.ml [hunk 8]
```diff
@@ -699,6 +713,40 @@ let rec unbox_float dbg =
     | Ctail e -> Ctail (unbox_float dbg e)
     | cmm -> Cop (Cload (Double, Immutable), [cmm], dbg))
 
+(* Vectors *)
+
+let box_vector dbg vi m c =
+  Cop (Calloc m, [alloc_boxedvector_header vi m dbg; c], dbg)
+
+let rec unbox_vector dbg vi =
+  let load = match vi with Primitive.Pvec128 -> Onetwentyeight in
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/cmm_helpers.ml:722]: removed this line

File: backend/amd64/proc.ml [hunk 9]
```diff
@@ -97,25 +98,64 @@ let float_reg_name =
...
+  and their ID ranges must not overlap.
+  
+  Here, both class 1 and class 2 refer to the set of `xmm` registers,
+  but are represented by different ID ranges. Class 1 is reserved for 
+  floating point values and class 2 is reserved for 128-bit SIMD vectors.
+  This allows us to preserve sizing information based on the register IDs.
+
+  The register allocators use the following two functions to check whether
+  allocating a register in one range must also consume an ID in another class.
+*)
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/proc.ml]: Switched back to using the same class for float and vec128

File: backend/amd64/proc.ml [hunk 10]
```diff
@@ -334,7 +417,7 @@ let destroyed_at_oper = function
   | Iop(Iextcall { alloc = false; }) -> destroyed_at_c_call
   | Iop(Iintop(Idiv | Imod)) | Iop(Iintop_imm((Idiv | Imod), _))
         -> [| rax; rdx |]
-  | Iop(Istore(Single, _, _)) -> [| rxmm15 |]
+  | Iop(Istore(Single, _, _)) -> [| rxmm15f; rxmm15v |]
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/proc.ml]: Switched back to one class

File: backend/cfg/cfg.ml [hunk 11]
```diff
@@ -504,7 +505,7 @@ let is_noop_move instr =
     (match instr.arg.(0).loc with
     | Unknown -> false
     | Reg _ | Stack _ -> Reg.same_loc instr.arg.(0) instr.res.(0))
-    && Proc.register_class instr.arg.(0) = Proc.register_class instr.res.(0)
+    && Proc.stack_slot_class_for instr.arg.(0) = Proc.stack_slot_class_for instr.res.(0)
```
Comment (gretay-js) [inline backend/cfg/cfg.ml:508]: It needs separate cases here for `Reg` to check `register_class` and `Stack` to check `stack_slot_class`.

File: middle_end/flambda2/numbers/numeric_types.ml [hunk 12]
```diff
@@ -250,3 +250,59 @@ module Int64 = struct
...
+
+  include T0
+  module Self = Container_types.Make (T0)
+  include Self
+
+  let zero = Array.init Width.size_in_int64s (fun _ -> 0L)
+
+  let to_int64_array t = t
+
+  let of_int64_array t = t
```
Comment (lthls) [inline middle_end/flambda2/numbers/numeric_types.ml:289]: I think this should include a length check. Or use an explicit name like `of_int64_array_unsafe`.

File: middle_end/flambda2/parser/print_fexpr.ml [hunk 13]
```diff
@@ -255,6 +257,7 @@ let const ppf (c : Fexpr.const) =
   | Naked_int32 i -> Format.fprintf ppf "%lil" i
   | Naked_int64 i -> Format.fprintf ppf "%LiL" i
   | Naked_nativeint i -> Format.fprintf ppf "%Lin" i
+  | Naked_vec128 { high; low } -> Format.fprintf ppf "%Ld:%Ld" high low
```
Comment (lthls) [inline middle_end/flambda2/parser/print_fexpr.ml:260]: Using `"%016Lx"` rather than `"%Ld"` might give a more consistent output (same for other printing functions later)

File: middle_end/flambda2/terms/code_size.ml [hunk 14]
```diff
@@ -120,14 +120,14 @@ let arith_conversion_size src dst =
 
 let unbox_number kind =
   match (kind : Flambda_kind.Boxable_number.t) with
-  | Naked_float -> 1 (* 1 load *)
+  | Naked_float | Naked_vec128 -> 1 (* 1 load *)
```
Comment (lthls) [inline middle_end/flambda2/terms/code_size.ml:123]: I'm not completely sure that it's accurate to count 128-bit loads as a single instruction, but I don't think it matters that much since all these numbers are rough guesses anyway.

File: middle_end/flambda2/numbers/numeric_types.ml [hunk 15]
```diff
@@ -250,3 +250,59 @@ module Int64 = struct
...
+    let compare l r = compare l r 0
+
+    let equal = Array.for_all2 Int64.equal
+
+    let hash v = Hashtbl.hash v
+
+    let print ppf t =
+      Format.pp_print_list
+        ~pp_sep:(fun ppf () -> Format.pp_print_char ppf ':')
+        Int64.print ppf (Array.to_list t)
```
Comment (lthls) [inline middle_end/flambda2/numbers/numeric_types.ml:278]: I think that we might want to print these numbers as fixed-length hexadecimal numbers (`Format.fprintf ppf "%016Lx"`)

**Review by lthls — APPROVED** [review]
Approving the middle-end and Cmm parts. I've left a few comments for areas that could be improved.

File: middle_end/flambda2/terms/code_size.ml (hunk 14 above)
Comment (TheNumbat) [reply middle_end/flambda2/terms/code_size.ml:123]: Yeah, it depends on how the backend implements 128-bit loads, but it's a single instruction on amd64 and arm64.

File: middle_end/flambda2/numbers/numeric_types.ml (hunk 15 above)
Comment (TheNumbat) [reply middle_end/flambda2/numbers/numeric_types.ml:278]: Changed to `%016Lx`

File: backend/amd64/proc.ml [hunk 16]
```diff
@@ -146,17 +146,15 @@ let hard_float_reg =
 let all_phys_regs =
   Array.append hard_int_reg hard_float_reg
 
-let phys_reg n =
-  if n < 100 then hard_int_reg.(n) else
-  if n < 200 then hard_float_reg.(n - 100)
-  else Misc.fatal_errorf "Register of unknown class (%d)" n
+let phys_reg ty n =
+  Reg.at_location ty (Reg n)
```
Comment (gretay-js) [inline backend/amd64/proc.ml:150]: I think we should not create a new one every time, preallocate them like before, but return by type. See comment in `Reg.reset`.

File: middle_end/flambda2/parser/print_fexpr.ml (hunk 13 above)
Comment (TheNumbat) [reply middle_end/flambda2/parser/print_fexpr.ml:260]: Changed to `%016Lx`

File: middle_end/flambda2/numbers/numeric_types.ml (hunk 12 above)
Comment (TheNumbat) [reply middle_end/flambda2/numbers/numeric_types.ml]: Added

File: backend/x86_ast.mli [hunk 17]
```diff
@@ -67,8 +68,9 @@ type reg64 =
 type reg8h =
   | AH | BH | CH | DH
 
+type registerf = XMMf of int | TOS | ST of int
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/x86_ast.mli]: Now removed

File: backend/amd64/proc.ml (hunk 16 above)
Comment (TheNumbat) [reply backend/amd64/proc.ml]: Now preallocated

File: backend/amd64/proc.ml [hunk 18]
```diff
@@ -133,18 +143,27 @@ let hard_float_reg =
...
   v
 
+let hard_vec128_reg =
+  let v = Array.make 16 Reg.dummy in
+  for i = 0 to 15 do v.(i) <- Reg.at_location Vec128 (Reg (100 + i)) done;
+  v
+
+(* We don't need to include vec128 regs here, as they use the same IDs as floats. *)
 let all_phys_regs =
   Array.append hard_int_reg hard_float_reg
```
Comment (gretay-js) [inline backend/amd64/proc.ml:153]: I am afraid we need to include it because of the way CSE uses the result of `destroyed_at_oper` as a key to `num_reg` which is of type Reg.Map in which the order is based on stamps.

Comment (TheNumbat) [reply backend/amd64/proc.ml:153]: Yeah, just ran into this - restored the destroyed_at behavior including both float and vec128

File: backend/amd64/proc.ml [hunk 19]
```diff
@@ -133,18 +144,28 @@ let hard_float_reg =
...
   v
 
+let hard_vec128_reg =
+  let v = Array.make 16 Reg.dummy in
+  for i = 0 to 15 do v.(i) <- Reg.at_location Vec128 (Reg (100 + i)) done;
+  v
+
 let all_phys_regs =
-  Array.append hard_int_reg hard_float_reg
+  Array.concat [hard_int_reg; hard_float_reg; hard_vec128_reg]
```
Comment (xclerc) [inline backend/amd64/proc.ml:153]: That does not feel right; the same register will 
appear in `hard_float_reg` and `hard_vec128_reg`
but the registers in `hard_int_reg` can be used
with different types. I am not positive this is
creating a problem, but the lack of homogeneity
gives me pause.

File: backend/amd64/proc.ml [hunk 20]
```diff
@@ -575,6 +627,15 @@ let init () =
   end else
     num_available_registers.(0) <- 13
 
+(* CR mslater for xclerc: is this ok? *)
+(* This is not always the same as [all_phys_regs], as some physical registers
+   may not be allocatable (e.g. rbp when frame pointers are enabled). *)
+let precolored_regs () =
```
Comment (xclerc) [inline backend/amd64/proc.ml:633]: I think it may break [that invariant](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/regalloc/regalloc_irc_state.ml#L536), but
it may be the invariant that needs updating.
Essentially, this is the same remark as for
`all_phys_regs`.

File: backend/amd64/proc.ml [hunk 21]
```diff
@@ -133,18 +144,28 @@ let hard_float_reg =
   for i = 0 to 15 do v.(i) <- Reg.at_location Float (Reg (100 + i)) done;
   v
 
+let hard_vec128_reg =
+  let v = Array.make 16 Reg.dummy in
+  for i = 0 to 15 do v.(i) <- Reg.at_location Vec128 (Reg (100 + i)) done;
```
Comment (xclerc) [inline backend/amd64/proc.ml:149]: Are we sure they should not have the same stamps
as the registers in `hard_float_reg`?

In concrete terms, do we think
`Reg.same hard_float_reg.(i) hard_vec128_reg.(i)`
should return `true` or `false`?

File: backend/regalloc/regalloc_invariants.ml [hunk 22]
```diff
@@ -146,7 +147,7 @@ let postcondition_layout : Cfg_with_layout.t -> unit =
       unit =
     match reg.Reg.loc with
     | Reg phys_reg ->
-      let phys_reg = Proc.phys_reg phys_reg in
+      let phys_reg = Proc.phys_reg reg.typ phys_reg in
```
Comment (xclerc) [inline backend/regalloc/regalloc_invariants.ml:150]: I think passing the type guarantees
the following condition cannot be
true.

File: ocaml/runtime/amd64.S [hunk 23]
```diff
@@ -347,47 +347,47 @@ LBL(caml_call_gc):
...
-        movsd   %xmm7, 7*8(%rsp)
-        movsd   %xmm8, 8*8(%rsp)
-        movsd   %xmm9, 9*8(%rsp)
-        movsd   %xmm10, 10*8(%rsp)
-        movsd   %xmm11, 11*8(%rsp)
-        movsd   %xmm12, 12*8(%rsp)
-        movsd   %xmm13, 13*8(%rsp)
-        movsd   %xmm14, 14*8(%rsp)
-        movsd   %xmm15, 15*8(%rsp)
+        subq    $(16*16), %rsp; CFI_ADJUST (16*16);
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply ocaml/runtime/amd64.S:350]: Yeah; I assume this change is fine since it's simple and movupd is supported in baseline x86_64

File: testsuite/tests/unboxed-primitive-args/test_common.h [hunk 24]
```diff
@@ -0,0 +1,50 @@
+/**************************************************************************/
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply testsuite/tests/unboxed-primitive-args/test_common.h:1]: Hmm, I just copied it from `ocaml/testsuite/tests/unboxed-primitive-args/text_commoh.h` - should I change it?

File: middle_end/flambda2/types/provers.ml [hunk 25]
```diff
@@ -874,17 +914,20 @@ let prove_physical_equality env t1 t2 =
     | Naked_nativeint (Ok s1), Naked_nativeint (Ok s2) ->
       let module IS = Targetint_32_64.Set in
       IS.is_empty (IS.inter (s1 :> IS.t) (s2 :> IS.t))
+    | Naked_vec128 (Ok s1), Naked_vec128 (Ok s2) ->
+      let module IS = Numeric_types.Vec128_by_bit_pattern.Set in
+      IS.is_empty (IS.inter (s1 :> IS.t) (s2 :> IS.t))
     | _, _ -> false
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply middle_end/flambda2/types/provers.ml:920]: Removed it myself

File: backend/amd64/proc.ml (hunk 21 above)
Comment (TheNumbat) [reply backend/amd64/proc.ml:149]: (After some discussion, we probably won't do this)

File: backend/regalloc/regalloc_invariants.ml (hunk 22 above)
Comment (TheNumbat) [reply backend/regalloc/regalloc_invariants.ml:150]: Good point, changed to check the stack slot class instead

File: backend/debug/reg_with_debug_info.ml [hunk 26]
```diff
@@ -99,23 +99,29 @@ let reg t = t.reg
...
      one is of class "Int" and another "Float" on amd64. [register_class] will
      be [Proc.register_class], but cannot be here, due to a circular
      dependency. *)
-  reg1.loc = reg2.loc && register_class reg1 = register_class reg2
+  reg1.loc = reg2.loc
+  &&
+  match reg1.loc with
+  | Reg _ -> register_class reg1 = register_class reg2
+  | Stack _ -> stack_class reg1 = stack_class reg2
+  | Unknown -> false
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/debug/reg_with_debug_info.ml]: Made this case an error

File: backend/amd64/proc.ml (hunk 19 above)
Comment (TheNumbat) [reply backend/amd64/proc.ml]: Added a flag guarding the generatin of Vec128 registers, and validated that we get the exact same assembly output for flambda-backend:main / this pr with the flag enabled and disabled. I'm pretty confident this is fine now; we could even remove the flag again.

File: backend/amd64/proc.ml [hunk 27]
```diff
@@ -622,6 +640,16 @@ let init () =
   end else
     num_available_registers.(0) <- 13
 
+(* This is not always the same as [all_phys_regs], as some physical registers
+   may not be allocatable (e.g. rbp when frame pointers are enabled). *)
+let precolored_regs =
+  let ints = Array.sub hard_int_reg 0 num_available_registers.(0) in
+  let floats = Array.sub hard_float_reg 0 num_available_registers.(1) in
+  let vec128s = Array.sub hard_vec128_reg 0 num_available_registers.(1) in
```
Comment (gretay-js) [inline backend/amd64/proc.ml:648]: is it possible that `precolored_regs` is computed before the first ever call to `Proc.init ()` ?
If so, `num_available_registers`  can be determined without running it, based on the configure option `fp`.

File: backend/amd64/proc.ml [hunk 28]
```diff
@@ -102,21 +102,32 @@ let num_register_classes = 2
...
-let register_class_tag c =
-  match c with
-  | 0 -> "i"
-  | 1 -> "f"
-  | c -> Misc.fatal_errorf "Unspecified register class %d" c
+  | Float -> 1
+  | Vec128 ->
+    if not !simd_regalloc_support then
+    Misc.fatal_error "SIMD register allocation is not enabled.";
+    1
```
Comment (gretay-js) [inline backend/amd64/proc.ml:109]: It's not clear to me that `register_class` can't be called on the result of `phys_reg` triggering the assert, but we will find out. 

It is unfortunate that `phys_reg` will readily return a `Reg.t` with `typ=vec128` even when simd_regalloc_support flag is off. I'd rather never create any `Reg.t` with that type when the flag is off. Is the only reason to creat them is to preallocate `hard_vec128_reg`?

Related to it, the distinction between the `precolored_regs` and `all_phys_regs` is not clear to me. They depend on the same compile-time flag. In fact, `precolored_regs` depends on `num_available_registers` that depens on `fp` with is a configure-time flag, so should be known statically. Concretely, `precolored_regs` can return a set not an array, using `Reg.set_of_array (all_phys_reg ())` and then `Reg.Set.remove rbp` from it if `fp` is true.

File: backend/amd64/proc.ml [hunk 29]
```diff
@@ -102,20 +102,31 @@ let num_register_classes = 2
...
+  | c -> Misc.fatal_errorf "Unspecified stack slot class %d" c
 
 let num_available_registers = [| 13; 16 |]
 
 let first_available_register = [| 0; 100 |]
 
 let register_name r =
-  if r < 100 then int_reg_name.(r) else float_reg_name.(r - 100)
+  if r < 100 then int_reg_name.(r)
+  else if r < 200 then float_reg_name.(r - 100)
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/amd64/proc.ml:138]: `register_name` shoudl take `typ` like `phys_reg`. it'll then be needed to distinguish xmm and ymm.

File: backend/amd64/emit.mlp [hunk 30]
```diff
@@ -104,12 +105,14 @@ let slot_offset loc cl =
...
-       | 2 -> n * 16
-       | 1 -> n * 8 + num_stack_slots.(2) * 16
-       | 0 -> n * 8 + num_stack_slots.(2) * 16 + num_stack_slots.(1) * 8
-       | _ -> Misc.fatal_error "Unknown register class")
+        (* Preserves original ordering (int -> float) *)
+        match cl with
+        | 2 -> n * 16
+        | 0 -> num_stack_slots.(2) * 16 + n * 8
+        | 1 -> num_stack_slots.(2) * 16 + num_stack_slots.(0) * 8 + n * 8
+        | _ -> Misc.fatal_error "Unknown register class")
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/amd64/emit.mlp:115]: I was wrong, this code is fine in terms of alignment, but it's weird. The reason to preserve original ordering of int and float is only for testing (to generate the same binary code).

File: backend/regalloc/regalloc_invariants.ml (hunk 22 above)
Comment (TheNumbat) [reply backend/regalloc/regalloc_invariants.ml:150]: Yeah, if we merge int/float this would be insufficient. Fixed

File: backend/amd64/proc.ml [hunk 31]
```diff
@@ -148,9 +159,10 @@ let hard_vec128_reg =
   for i = 0 to 15 do v.(i) <- Reg.at_location Vec128 (Reg (100 + i)) done;
   v
 
-(* We don't need to include vec128 regs here, as they use the same IDs as floats. *)
 let all_phys_regs =
-  Array.append hard_int_reg hard_float_reg
+  let basic_regs = Array.append hard_int_reg hard_float_reg in
+  let simd_regs = Array.append basic_regs hard_vec128_reg in
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/proc.ml:164]: Moved to top level (though it was already staged here)

File: backend/amd64/proc.ml [hunk 32]
```diff
@@ -344,48 +362,48 @@ let regs_are_volatile _rs = false
...
+  fun () -> if !simd_regalloc_support then simd_regs else basic_regs
+
+let destroyed_at_alloc_or_poll () =
   if X86_proc.use_plt then
     destroyed_by_plt_stub
   else
     [| r11 |]
 
-let destroyed_at_pushtrap =
+let destroyed_at_pushtrap () =
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/proc.ml:386]: I had just added that for consistency, but it doesn't - removed

File: backend/amd64/proc.ml (hunk 27 above)
Comment (TheNumbat) [reply backend/amd64/proc.ml:648]: Edited to check fp instead of relying on init

File: backend/amd64/proc.ml (hunk 28 above)
Comment (TheNumbat) [reply backend/amd64/proc.ml:109]: Made phys_reg check the flag, which means `destroyed_at_c_call` now uses `hard_vec128_reg` directly. I think the result is slightly better.

File: backend/regalloc/regalloc_invariants.ml (hunk 22 above)
Comment (xclerc) [reply backend/regalloc/regalloc_invariants.ml:150]: I am still not convinced: we used to
check that `reg` and `phys_reg`, taken
independently, were compatible.
We now check they are compatible,
but information from `reg` is used to
build `phys_reg`.

So I am unsure we are really checking
anything. And the condition is probably
redundant: if registers are in the same
class, aren't they necessarily in the same
stack class?

Perhaps the check can no longer exist,
given the new `Proc`.

File: backend/amd64/proc.ml [hunk 33]
```diff
@@ -133,10 +133,22 @@ let num_available_registers = [| 13; 16 |]
...
+    float_reg_name.(r - first_available_register.(1))
+  | Vec128 ->
+    if not !simd_regalloc_support then
+    Misc.fatal_error "SIMD register allocation is not enabled.";
+    float_reg_name.(r - first_available_register.(1))
+
+let register_name_lossy idx =
+  if idx < 100 then int_reg_name.(idx)
+  else if idx < 200 then float_reg_name.(idx - 100)
+  else Misc.fatal_errorf "Unknown register ID %d" idx
```
Comment (gretay-js) [inline backend/amd64/proc.ml:151]: We should replace the magic 100 and 200 constants with `typ` as in register_name, and we should keep this "lossy" business inside the validator. Perhaps the right thing to do is to record `typ` in `Description.Register.t`.

File: backend/x86_binary_emitter.ml [hunk 34]
```diff
@@ -438,7 +436,7 @@ let emit_mod_rm_reg b rex opcodes rm reg =
       emit_rex b (rex lor rex_of_reg16 reg16 lor rexr_reg reg lor rexb_rm rm);
       buf_opcodes b opcodes;
       buf_int8 b (mod_rm_reg 0b11 rm reg)
-  | Regf rm ->
+  | Regf (XMM _ as rm) ->
```
Comment (gretay-js) [inline backend/x86_binary_emitter.ml:439]: Why do we need to match on XMM here? I don't think it will help when other constructors are added to `regf` .

File: backend/x86_binary_emitter.ml [hunk 35]
```diff
@@ -1398,20 +1403,17 @@ let emit_FLDCW b = function
   | (Mem _ | Mem64_RIP _) as rm -> emit_mod_rm_reg b no_rex [ 0xD9 ] rm 5
   | _ -> assert false
 
-let emit_FXCH b = function
-  | Regf (ST i) -> buf_opcodes b [ 0xD9; 0xC8 + i ]
+let emit_FXCH _b = function
```
Comment (gretay-js) [inline backend/x86_binary_emitter.ml:1406]: Can these instructions be removed too? If so, can you please make a separate PR for the changes in registerf (removal of TOS and ST and the corresponding instructions)?

File: backend/amd64/emit.mlp [hunk 36]
```diff
@@ -49,7 +49,8 @@ let float_reg_name = Array.init 16 (fun i -> XMM i)
 
 let register_name r =
   if r < 100 then Reg64 (int_reg_name.(r))
-  else Regf (float_reg_name.(r - 100))
+  else if r < 200 then Regf (float_reg_name.(r - 100))
+  else Misc.fatal_errorf "Bad register id %d" r
```
Comment (gretay-js) [inline backend/amd64/emit.mlp:53]: I think this should also use `typ` instead of magic 100 and 200 constants.

File: backend/regalloc/regalloc_stack_slots.ml
Comment (xclerc) [inline backend/regalloc/regalloc_stack_slots.ml:1]: I think various comments are no longer in
sync with the code; something along the
lines of `s/register class/stack class/g`
would need to happen.

File: backend/amd64/proc.ml (hunk 33 above)
Comment (TheNumbat) [reply backend/amd64/proc.ml:151]: Refactored printing in validator to not require this

File: backend/regalloc/regalloc_invariants.ml (hunk 22 above)
Comment (TheNumbat) [reply backend/regalloc/regalloc_invariants.ml:150]: Updated to just run `phys_reg`

File: backend/amd64/proc.ml (hunk 31 above)
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/proc.ml]: Refactored again

File: backend/amd64/proc.ml (hunk 28 above)
Comment (TheNumbat) [reply backend/amd64/proc.ml:109]: Refactored again

File: backend/amd64/proc.ml [hunk 37]
```diff
@@ -643,12 +660,17 @@ let init () =
...
+  let basic_regs = Array.append hard_int_reg hard_float_reg in
+  let basic_regs_no_fp = Array.append ints_no_fp hard_float_reg in
+  let basic_and_simd_regs = Array.append basic_regs hard_vec128_reg in
+  let basic_and_simd_regs_no_fp = Array.append basic_regs_no_fp hard_vec128_reg in
+  fun () ->
+    match fp, !simd_regalloc_support with
+    | false, false -> basic_regs
+    | true, false -> basic_regs_no_fp
+    | false, true -> basic_and_simd_regs
+    | true, true -> basic_and_simd_regs_no_fp
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/proc.ml]: Refactored again

File: backend/x86_binary_emitter.ml (hunk 35 above)
Comment (TheNumbat) [reply backend/x86_binary_emitter.ml:1408]: PR #1555

File: backend/regalloc/regalloc_invariants.ml [hunk 38]
```diff
@@ -145,13 +146,13 @@ let postcondition_layout : Cfg_with_layout.t -> unit =
   let register_classes_must_be_consistent (id : Instruction.id) (reg : Reg.t) :
       unit =
     match reg.Reg.loc with
-    | Reg phys_reg ->
-      let phys_reg = Proc.phys_reg phys_reg in
-      if not (same_reg_class reg phys_reg)
-      then
+    | Reg phys_reg -> (
+      try Proc.phys_reg reg.typ phys_reg |> ignore
```
Comment (xclerc) [inline backend/regalloc/regalloc_invariants.ml:150]:
```suggestion
      try let _ : Reg.t = Proc.phys_reg reg.typ phys_reg in ()
```

**Review by gretay-js — APPROVED** [review]
It's ready from my point of view, once the last minor comments are addressed.  Thanks @TheNumbat  for all the work!

**Comment by gretay-js** [comment]
This PR breaks several properties that the backend may be implicitly relying on for correctness, mainly during register allocation. I'll try to describe them, based on an offline discussion with @xclerc.

- A register class can now represent temporaries of different sizes that are determined by `Reg.typ`. This is handled by introduction of stack slot classes for different sizes of slots, and checking tge type, not only register class. 
- Introduction of hard registers with the same register index and name but different types (`hard_vec128_reg` in addition to  `hard_float_regs` in `Proc`). The new registers with vec128 type will be created even if `simd_regalloc_support` flag is off, making the stamps incomparable. However, all uses of the new hard registers are guarded by the flag. 
- When the flag is off, the generated assembly will be identical to what we get before the PR (tested on large code base), and effectively no difference in build times and memory. The internal representation is nearly the same, except for stamps, as mentioned above.
- When the flag is on, the names in `all_phys_reg` are no longer unique. 
- Coaleasing based on register class may attempt to coalease registers of different sizes.  This is fine for now even when the flag is on, because the only case uses the same register names (i.e., float and vec128 both result in xmm registers). It will need to be addressed for ymm and zmm, if we go ahead with the current plan of representing them using the same register classes (because they use the same physical space).
- Validator does not support subregs. Implement the check for overlapping registers from the paper.
- Compiler never generates `Imove` between registers float and vec128 registers.

**Comment by xclerc** [comment]
About coalescing, one way to ensure there will not
be any problems is to be sure `Move` is not used
between registers of different types. I understand
it is Max's intention for the new types, and I think
I would go as far as to say that `Emit.move` should
only accept `src` and `dst` with the same type.

File: backend/amd64/emit.mlp [hunk 39]
```diff
@@ -718,18 +718,24 @@ let emit_global_label s =
 let move (src : Reg.t) (dst : Reg.t) =
   if src.loc <> dst.loc then
     begin match src.typ, src.loc, dst.typ, dst.loc with
-    | Float, Reg.Reg _, Float, Reg.Reg _
-    | Vec128, Reg.Reg _, Vec128, Reg.Reg _ -> I.movapd (reg src) (reg dst)
+    | (Float | Vec128), Reg.Reg _, (Float | Vec128), Reg.Reg _ ->
```
Comment (gretay-js) [inline backend/amd64/emit.mlp:721]: This should stay as before, only `float,float` and `vec128,vec128` pairs are allowed.

File: backend/amd64/emit.mlp [hunk 40]
```diff
@@ -718,18 +718,24 @@ let emit_global_label s =
...
+      I.movapd (reg src) (reg dst)
     | Vec128, _, Vec128, _ ->
       I.movupd (reg src) (reg dst)
```
Comment (gretay-js) [inline backend/amd64/emit.mlp]: Can this use `movapd`? The stack slots are always aligned.

File: backend/amd64/emit.mlp [hunk 41]
```diff
@@ -718,18 +718,24 @@ let emit_global_label s =
...
-    | Float, _, Float, _ -> I.movsd (reg src) (reg dst)
-    | Float, _, Int, _
-    | Int, _, Float, _ -> I.movq (reg src) (reg dst)
+    | Float, _, Float, _ ->
+      I.movsd (reg src) (reg dst)
+    | Float, _, Int, _ | Int, _, Float, _ ->
+      I.movq (reg src) (reg dst)
```
Comment (gretay-js) [inline backend/amd64/emit.mlp:728]: Please can you add a CR here while you are at it about bits_of_float / float_of_bits and the plan to get rid of this case soon, using a separate `Ispecific`.

**Review by gretay-js — APPROVED** [review]
